### ISSUE TITLE
BACKUP=REQUESTRESTORE does not automatically include mount.nfs so cannot mount nfs share to restore backup.tar.gz

### ISSUE THREAD

**Issue by Testicule** [rear/rear#1911]
* ReaR version ("/usr/sbin/rear -V"):
             Relax-and-Recover 2.4 / 2018-06-21
* OS version:
             SUSE Linux Enterprise Server 11 SP4
* Hardware:
             IBM SAP HANA appliance

* Description of the issue:

backup.tar.gz does not fit in 4GB ISO, so i have a small rescue ISO, and backup.tar.gz is saved to nfs share.  When testing recovery, i set BACKUP=REQUESTRESTORE, then do "rear recover". This successfully recreated disk partitions etc. I then setup network.

Then i attempt to mount the  nfs share:
    mount -t nfs -o nfsvers=3,ro   IPADDR:/export/linux_images   /mnt/nfs_server

Result, is epic failure:
   "mount: wrong fs type, bad option, bad superblock on blah, 
        missing codepage or helper program, or other error
        (for several filesystems (e.g. nfs, cifs) you might
        need a /sbin/mount.<type> helper program)
        blah blah blah"

i have done a find, and the rescue image has an nfs kernel module. But there is no mount.nfs

How on earth are we expected to restore from nfs network share, with no mount.nfs?!

* Work-around, if any:
None

**Comment by gdha** [comment]
Just add the following to the `local.conf` file;
````
PROGS=( "${PROGS[@]}" showmount mount.nfs umount.nfs )
MODULES=( "${MODULES[@]}" nfs )
PRE_RECOVERY_SCRIPT="systemctl start rpcbind.target || rpcbind &"
````

**Comment by jsmeix** [comment]
With our current ReaR on our planet earth ;-)
you are expected to do some trial and error legwork
together with reading through various documentation
to build up your particular disaster recovery procedure
that works in your particular case, see in particular
the section "Inappropriate expectations" at
https://en.opensuse.org/SDB:Disaster_Recovery

Regarding your
<pre>
backup.tar.gz does not fit in 4GB ISO
</pre>
you may have a look at
https://github.com/rear/rear/blob/master/doc/user-guide/11-multiple-backups.adoc
because I assume a backup of only the files of your basic system
would fit in a 4GB ISO so that this way cou could use a clean
standard ReaR setup where ReaR is used only for what it is menat
to be used: to recreate the basic operating system, cf.
https://en.opensuse.org/SDB:Disaster_Recovery

If you really need a huge all-in-one backup that forces you
as a consequence to restore manually via `REQUESTRESTORE`
then you cannot expect that during "rear mkrescue/mkbackup"
ReaR magically (there is not actually working magic on planet earth ;-)
knows what programs, libraries, and other files you need
to do your kind of manual backup restore so that
as a further consequence you also need to tell ReaR
what programs, libraries, and other files you need
to do your kind of manual backup restore via
ReaR config variables like `COPY_AS_IS`,
`REQUIRED_PROGS`, and `LIBS`.

In general regarding ReaR config variables see the `default.conf` file
e.g. our current one of the latest ReaR upstream code at
https://raw.githubusercontent.com/rear/rear/master/usr/share/rear/conf/default.conf

**Comment by Testicule** [comment]
Thank you for the suggestions. I hear what you're saying, but:
  1) showmount and nfs kernel module were included automatically.
  2) let's pls just ensure we also include mount.nfs by default. 

It is not an unusual scenario to want to restore from an nfs share.

People do not always have the luxury of being able to test and re-test as much as they would like. Eg running rear on a production box, even with "nice" there may be concerns about the additional load, and it may be hard to get permission to re-run backups. So it would be nice if the default configuration helped get a good backup the first time. 

Obviously, we want things to go smoothly, where possible. 

And I think we can do without this problem tripping people up.

**Comment by Testicule** [comment]
Incidentally, I have also had problems with putting my backup.tar.gz into a nice new VMWare virtual disk, and attaching that to a VM, to test recovery. 

I was unable to mount an xfs filesystem, which was incredibly frustrating...
Then i tried ext4. No good.
Eventually had to put my backup.tar.gz on an ext3 filesystem, and attach it.

So, please. Can we reconsider putting a few of these things in the recovery iso, by default?
Xfs is now a very, very common filesystem. It's the recommended fs for many puproses.
And ext4. 

Surely we can include the ability to mount nfs, xfs, and ext4 by default in every rescue image (as long as the host system can do so).

I know users can add these things to their rescue images manually - but these are really, really basic tools. And people will get themselves into situations they did not completely foresee, when doing recoveries. Disks will be unavailable, DVD drives can fail, but if you can setup networking and use nfs, and other filesystems, you can work around these issues and succeed.

**Comment by Testicule** [comment]
I mean, I'm booted onto a rescue disk created with the standard rear on RHEL7, and i actually have mkfs.xfs, so I can create an xfs filesystem... but i cannot mount it. So that's a bit ridiculous.

We really need ReaR to have these mount programs, by default, please.

**Comment by Testicule** [comment]
When i create the backups, I AM using NETFS.

I set REQUESTRESTORE for several reasons:
1) So i could skip all the network interface configuration, and start with just getting ReaR able to recreate all the filesystems. This has required some work, because I am testing the restore to an ESX host with smaller disk than the source host. The main system fits on the smaller disk, the original host had a large sda4, and i can get away with shrinking that right down. So I'm having to tweak the diskrestore.sh and i have successfully got ReaR to get the disks done, and filesystems recreated without too much fuss.

2) Then - manually setup my networking, get NFS share mounted and extract backup tarball.

I have to be very careful with the restore, ensuring ReaR does not re-instate the original IP addresses - because I do not want my recovery testing to cause IP conflicts with the production box which is still running. If i cause network conflicts and disrupt anything on the prod box and have to reboot it for example, it takes 45 minutes just to reboot as it's a SAP HANA monster with 287 processors, and 2TB of RAM.

If i had mount.nfs, I'd have sailed past step2 without any grief.

Still, I have now managed to recreate the rescue disk with mount.nfs, thanks to your suggestions - but I was worried I wouldn't be able to do it for some time, as with prod servers, there are restrictions in what can be run on them willy-nilly. But i got it done.

ReaR will be a more robust/versatile tool with mount.nfs always included. It's only small, it's worth it. And same goes for mounting xfs & ext4. With xfs & ext4, IIRC, the problem was missing kernel modules, so i should have them now i'm using MODULES=( 'all_modules' ). But, it would be so much better if these were included by default.

Nfs, xfs, and ext4 should always be available by default - if the host system supported them, pls include them in the rescue iso.

**Comment by jsmeix** [comment]
The current ReaR works as intended and this cannot be changed
at least not in the near future - except there is someone who
contributes the needed changes to us at ReaR upstream.

The current "rear mkrescue/mkbackup" creates a recovery system
that is rather minimal to save space where the recovery system
ISO image is stored (e.g. on a NFS server).

Personally I was hit often enough by missing "usually expected stuff"
in the currently rather minimal ReaR recovery system which is in
particular bad because in parctice it is usually a dead end
when one notices in the running recovery system that
something is missing, e.g. see
https://github.com/rear/rear/issues/1202

Therefore with current ReaR you cannot just run "rear mkbackup"
on your production server and relax, see the sections
"No disaster recovery without testing and continuous validation"
and
"Let's face it: Deployment via the recovery installer is a must" in
https://en.opensuse.org/SDB:Disaster_Recovery

In current ReaR the recovery system is specific for the one particular
system where "rear mkrescue/mkbackup" was run), see the
"Essentials about disaster recovery with Relax-and-Recover presentation PDF"
that is linked in the "See also" section in
ttps://en.opensuse.org/SDB:Disaster_Recovery

When you have hundreds of servers it makes a noticeable
difference how much total space is needed where all those
hundreds of ReaR recovery systems are stored.

My own plan for an unspecified future (as time permits but usually
I have no time left for long term useful enhancements because
usually any real future work is blocked by the day-to-day stuff)
is to enhance ReaR so that it can alternatively produce a generally
usable full featured recovery system that runs basically on any hardware
just like the usual Linux installation systems behave that you get
as installation medium from the various Linux distributions,
cf. "My ultimate goal" at
https://github.com/rear/rear/issues/1085

For a somewhat related issue you may have a look at
https://github.com/rear/rear/issues/841
where RECOVERY_UPDATE_URL is intended to mitigate the needed
storage space when one has many servers with identical hardware
so that one same hardware specific ReaR recovery system would work
because the only differences are some ReaR configuration files.

**Comment by jsmeix** [comment]
Regardless that I closed this issue we can
continue to add comments here:

Regarding "restore to an ... host with smaller disk":
Careful!
I think in practice you must use at least ReaR 2.4
because before version 2.4 the old behaviour may result
unexpected changes of arbitrary partitions on a disk, see
the ReaR 2.4 release notes e.g. on
http://relax-and-recover.org/documentation/release-notes-2-4

Regarding things like 'mkfs.whatever' not included by default
in the ReaR recovery system:
This again works as intended by the current ReaR to keep
its recovery system rather minimal.
In particular regarding XFS cf. the mail thread about
"Migrating from ReiserFS to a differen Filesystem"
on the rear-users mailing list archive e.g. at
http://lists.relax-and-recover.org/pipermail/rear-users/2018-April/003547.html
where I wrote in particular (excerpt)
<pre>
A crucial point to make this "just work" with ReaR is to have
already some (tiny) mounted XFS partition on the system when
"rear mkbackup" runs because that mounted XFS partition triggers
the automatisms in ReaR to get all what is needed for XFS setup
included in the ReaR recovery system (without that things
like 'mkfs.xfs' would "just fail" during "rear recover").
</pre>
